src/drove.go

```diff
@@ -512,35 +534,33 @@ func nginxPlus() error {
 			"upstreams": newFormattedServers,
 		}).Info("nginx upstreams")
 
-		logger.WithFields(logrus.Fields{
-			"nginx": config.Nginxplusapiaddr,
-		}).Info("endpoint")
-
-		endpoint := "http://" + config.Nginxplusapiaddr + "/api"
-
-		tr := &http.Transport{
-			MaxIdleConns:       30,
-			DisableCompression: true,
-		}
-
-		client := &http.Client{Transport: tr}
-		nginxClient, error := nplus.NewNginxClient(endpoint,nplus.WithHTTPClient(client),nplus.WithAPIVersion(
-			8))
-		if error != nil {
-			logger.WithFields(logrus.Fields{
-				"error": error,
-			}).Error("unable to make call to nginx plus")
-			return error
-		}
 		upstreamtocheck := app.Vhost
 		var finalformattedServers []nplus.UpstreamServer
 
 		for _, server := range newFormattedServers {
 			formattedServer := nplus.UpstreamServer{Server: server, MaxFails: config.MaxFailsUpstream, FailTimeout: config.FailTimeoutUpstream, SlowStart: config.SlowStartUpstream}
 			finalformattedServers = append(finalformattedServers, formattedServer)
 		}
+		// If upstream has no servers, UpdateHTTPServers returns error as in-line GetHTTPServers returns error. server ID 0 needs to be explicitly initiated by a PATCH
                 err := nginxClient.CheckIfUpstreamExists(upstreamtocheck)
-                if err != nil {
+		if err != nil {
+			// First add atleast one server to initialise upstream to support UpdateHTTPServers
+			logger.WithFields(logrus.Fields{
+				"Adding fresh upstream for": upstreamtocheck,
+			}).Info("Adding first server for upstream")
+			//Adding first server for server ID 0. ID 0 needs to be updated if state file is resurrected when a vhost gets resurrected. Create ID 0 otherwise.
+			error := nginxClient.UpdateHTTPServer(upstreamtocheck, finalformattedServers[0])
+
+			// Now upstream should have servers, update earlier state to let UpdateHTTPServers take over
+			//But wait from some time for nginx to actually update it's state. Consecutive calls would still return a 404 if you don't wait long enough
+			if error != nil {
+				time.Sleep(100 * time.Millisecond)
```

Comment:
This should be a retry loop instead of static wait

Reply:
Added wait for loop with a context timeout to avoid waiting forever